### PR TITLE
search: show a "Did you mean" link for invalid regexes

### PR DESCRIPTION
**Pull request by ijsnow**
Add a "Did you mean" link for invalid regexes in search queries.

Closes https://github.com/sourcegraph/sourcegraph/issues/2772.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ All notable changes to Sourcegraph are documented in this file.
   To enable, open the user menu in the top right and make sure the theme dropdown is set to "System".
   This is currently supported on macOS Mojave with Safari Technology Preview 68 and later.
 - The `github.exclude` setting was added to the [GitHub external service config](https://docs.sourcegraph.com/admin/external_service/github#configuration) to allow excluding repositories yielded by `github.repos` or `github.repositoryQuery` from being synced.
+- We now suggest fixes for common invalid regexes in search queries.
```
Comment (ijt) [inline CHANGELOG.md]: We now provide "Did you mean...?" links when search queries fail to parse as regular expressions.

File: CHANGELOG.md [hunk 2]
```diff
@@ -813,8 +814,8 @@ You may now optionally provide the SAML Identity Provider metadata XML file cont
 
 When using `auth.provider == "builtin"`, two new important changes mean that a Sourcegraph server will be locked down and only accessible to users who are invited by an admin user (previously, we advised users to place their own auth proxy in front of Sourcegraph servers).
 
-1.  When `auth.provider == "builtin"` Sourcegraph will now by default require an admin to invite users instead of allowing anyone who can visit the site to sign up. Set `auth.allowSignup == true` to retain the old behavior of allowing anyone who can access the site to signup.
-2.  When `auth.provider == "builtin"`, Sourcegraph will now respects a new `auth.public` site configuration option (default value: `false`). When `auth.public == false`, Sourcegraph will not allow anyone to access the site unless they have an account and are signed in.
+1. When `auth.provider == "builtin"` Sourcegraph will now by default require an admin to invite users instead of allowing anyone who can visit the site to sign up. Set `auth.allowSignup == true` to retain the old behavior of allowing anyone who can access the site to signup.
```
Comment (ijt) [inline CHANGELOG.md]: This chunk seems unrelated. Please revert.

File: cmd/frontend/internal/pkg/search/query/types/check.go [hunk 3]
```diff
@@ -44,16 +44,29 @@ func (c *Config) Check(query *syntax.Query) (*Query, error) {
...
+	fixedExprs := []string{}
+	errCount := 0
+
 	for _, expr := range query.Expr {
+		fx := expr.String()
 		field, fieldType, value, err := c.checkExpr(expr)
 		if err != nil {
-			return nil, err
+			errCount++
+			fx = `"` + fx + `"`
```
Comment (ijt) [inline cmd/frontend/internal/pkg/search/query/types/check.go]: Let's make this more specific:

```suggestion
if err != nil {
	if expr.ValueType != syntax.TokenPattern {
		return err
	}
	errCount++
	fx = `"` + fx + `"`
}
```

File: cmd/frontend/internal/pkg/search/query/types/check.go [hunk 4]
```diff
@@ -44,16 +44,29 @@ func (c *Config) Check(query *syntax.Query) (*Query, error) {
...
+		fixedExprs = append(fixedExprs, fx)
 		if fieldType.Singular && len(checkedQuery.Fields[field]) >= 1 {
 			return nil, &TypeError{Pos: expr.Pos, Err: fmt.Errorf("field %q may not be used more than once", field)}
 		}
 		checkedQuery.Fields[field] = append(checkedQuery.Fields[field], value)
 	}
+
+	if errCount > 0 {
+		s := strings.Join(fixedExprs, " ")
+		return nil, fmt.Errorf("Error encountered while parsing regular expression. Did you mean `%s`?", s)
```
Comment (ijt) [inline cmd/frontend/internal/pkg/search/query/types/check.go]: Also store up all the error strings and include them here in case it's something more subtle where the user actually was writing a real regex and got something wrong.

File: CHANGELOG.md (hunk 2 above)
Comment (ijt) [reply CHANGELOG.md]: Hmm, this should still be reverted.

File: web/src/search/results/ResultsError.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            </div>
+        )
+    }
+
+    private renderMessage(): React.ReactNode {
+        const {
+            error: { message },
+        } = this.props
+
+        // TODO: Send more robust error response from backend to prevent the need to string match.
```
Comment (felixfbecker) [inline web/src/search/results/ResultsError.tsx]: Yeah this is really brittle. How do you plan to solve it?

Comment (felixfbecker) [reply web/src/search/results/ResultsError.tsx]: Am I understanding correctly that this is a different mechanism than this?

image

If yes, why?

**Comment by sqs** [comment]
Closing, see https://github.com/sourcegraph/sourcegraph/issues/2125#issuecomment-475514680
